language/diem-framework/modules/NetworkIdentity.move

```diff
@@ -0,0 +1,143 @@
+        *&borrow_global<NetworkIdentity>(account_addr).identities
+    }
+
+    spec get {
+    }
+
+    /// Update and create if not exist `NetworkIdentity`
+    public fun add_identities(account: &signer, to_add: vector<vector<u8>>) acquires NetworkIdentity {
+        if (!exists<NetworkIdentity>(Signer::address_of(account))) {
+            initialize_network_identity(account);
+        };
+        let identity = borrow_global_mut<NetworkIdentity>(Signer::address_of(account));
+        let identities = &mut identity.identities;
+        let i = 0;
+        let has_change = false;
+        while (i < Vector::length(&to_add)) {
+           has_change = has_change || add_identity(identities, *Vector::borrow(&to_add, i));
+           i = i + 1;
+        };
+
+        assert(Vector::length(identities) <= MAX_ADDR_IDENTITIES, Errors::limit_exceeded(ENET_ID_LIMIT_EXCEEDED));
+
+        if (has_change) {
+            Event::emit_event(&mut identity.identity_change_events, NetworkIdentityChangeNotification {
+                identities: *&identity.identities,
+                time_rotated_seconds: DiemTimestamp::now_seconds(),
+            });
+        }
+    }
+
+    spec add_identities {
+        let account_addr = Signer::spec_address_of(account);
+        ensures exists<NetworkIdentity>(account_addr);
+    }
+
+    /// Adds an identity and returns true if a change was made
+    fun add_identity(identities: &mut vector<vector<u8>>, to_add: vector<u8>): bool {
+        if (!Vector::contains(identities, &to_add)) {
+            Vector::push_back(identities, to_add);
+            true
+        } else {
+            false
+        }
+    }
+
+    spec add_identity {
+        ensures contains<vector<u8>>(identities, to_add);
+    }
+
+    /// Remove `NetworkIdentity`, skipping if it doesn't exist
+    public fun remove_identities(account: &signer, to_remove: vector<vector<u8>>) acquires NetworkIdentity {
+        if (!exists<NetworkIdentity>(Signer::address_of(account))) {
+            return
+        };
+        let identity = borrow_global_mut<NetworkIdentity>(Signer::address_of(account));
+        let identities = &mut identity.identities;
+
+        let i = 0;
+        let has_change = false;
+        while (i < Vector::length(&to_remove)) {
+           has_change = has_change || remove_identity(identities, *Vector::borrow(&to_remove, i));
+           i = i + 1;
+        };
+
+        if (has_change) {
+            Event::emit_event(&mut identity.identity_change_events, NetworkIdentityChangeNotification {
+                identities: *&identity.identities,
+                time_rotated_seconds: DiemTimestamp::now_seconds(),
+            });
+        };
+    }
+
+    /// Removes an identity and returns true if a change was made
+    fun remove_identity(identities: &mut vector<vector<u8>>, to_remove: vector<u8>): bool {
+        let removed = false;
+
+        loop {
```

Comment:
Do you need a `loop` here? The set property that is enforced by the `add_identity` function should mean that only one iteration of this loop is ever done. If so, I would prefer if the `loop` was removed here since having unbounded looping in the Diem Framework could cause issues if for some reason it gets into a state where an infinite loop could happen.

Reply:
I added the loop due to trying to get the spec to pass, as at this point it doesn't know that it is a set (at least what I know from the spec language).

Do we actually need a loop, given we provide it as an invariant above? No.